File: day7/day7.rb
require_relative 'elf.rb'
require_relative 'task.rb'

# raw = %q(Step C must be finished before step A can begin.
# Step C must be finished before step F can begin.
# Step A must be finished before step B can begin.
# Step A must be finished before step D can begin.
# Step B must be finished before step E can begin.
# Step D must be finished before step E can begin.
# Step F must be finished before step E can begin.)
raw = File.read("input.txt")

requirements = raw.scan(/Step\s(.).+step\s(.)\s/)

steps = {}
requirements.each do |rule|
  steps[rule[0]] ||= Task.new(rule[0])
  steps[rule[1]] ||= Task.new(rule[1])
  steps[rule[1]].requirements << steps[rule[0]]
end

sequence = []
while sequence.length < steps.length
  valid_steps = steps.select {|k,v| !v.finished? && v.can_run?}
  next_step = valid_steps.min[1]
  sequence << next_step.name
  next_step.status = :finished
end

p sequence.join

# Have to reset all the steps for part 2
steps.each do |k, v|
  v.status = :start
end

seconds = 0
elves = Array.new(5){Elf.new}
loop do
  elves.each do |e|
    if e.tasked?
      e.task.seconds -= 1
      if e.task.finished?
        e.reset
      end
    end
  end

  valid_steps = steps.select {|k, v| v.can_run?}.values.sort
  elves.find_all {|e| !e.tasked?}.each_with_index do |elf, i|
    if i >= valid_steps.length
      break
    else
      elf.set(valid_steps[i])
    end
  end

  break if steps.none? {|k, v| !v.finished?}
  seconds += 1
end

p seconds

File: day7/day7.js
// var input = `Step C must be finished before step A can begin.
// Step C must be finished before step F can begin.
// Step A must be finished before step B can begin.
// Step A must be finished before step D can begin.
// Step B must be finished before step E can begin.
// Step D must be finished before step E can begin.
// Step F must be finished before step E can begin.`
var input = raw
  .split("\n")
  .map(x => x.match(/Step (\D) must be finished before step (\D) can begin./).slice(1));

var steps = {};
input.forEach((rule) => {
  rule.forEach((step) => {
    if(steps[step] == undefined){
      steps[step] = {allows: [], requires: [], completed: false}
    }
  });
  steps[rule[0]].allows.push(rule[1]);
  steps[rule[1]].requires.push(rule[0]);
});

var assigned = [];
var positions = [];
var i = 0;
while(assigned.length < Object.keys(steps).length){
  positions.push([]);
  var newly_assigned = []
  Object.keys(steps).forEach((key) => {
    if(steps[key].requires.every((n) => assigned.indexOf(n) > -1) && !assigned.includes(key)){
      positions[i].push(key);
      newly_assigned.push(key);
    }
  });
  positions[i].sort();
  assigned = assigned.concat(newly_assigned);
  i++;
}

var canvas_width = 800;
var canvas_height = 500;
var column_width = canvas_width/positions.length;
var offset = canvas_height * 0.05;
for(var i = 0; i < positions.length; i++){
  var x = column_width * i + 10;
  var row_height = (canvas_height)/(positions[i].length + 1) + offset;
  for(var j = 0; j < positions[i].length; j++){
    var key = positions[i][j];
    steps[key].x = x;
    steps[key].y = (row_height) * (j+1) - ((positions[i].length + 1) * offset/2);
  }
}

function setup() {
  createCanvas(canvas_width, canvas_height);

  var id = setInterval(() => {
    var allowed = Object.keys(steps)
      .filter((k) => steps[k].requires.every((n) => sequence.includes(n)) && !sequence.includes(k))
      .sort()[0];
    sequence += allowed;
    steps[allowed].completed = true;
    if(sequence.length == Object.keys(steps).length){
      console.log("We're done! Yay!");
      clearInterval(id);
    }
  }, 1000);
}

var sequence = "";
function draw() {
  background(255);

  strokeWeight(1);
  Object.keys(steps).forEach(key => {
    var start = steps[key]
    start.requires.forEach(r => {
      var end = steps[r];
      if(!start.completed && end.completed){
        stroke(0);
      }else if(start.completed && end.completed){
        stroke(240);
      }else{
        stroke(150);
      }
      line(start.x, start.y, end.x, end.y);
    });
  });

  textSize(12);
  stroke(0);
  strokeWeight(1);
  Object.keys(steps).forEach(key => {
    var step = steps[key]
    if(step.completed){
      fill(255);
    }else if(step.requires.every((n) => sequence.includes(n))){
      fill(200);
    }else{
      fill(120);
    }
    if(step.x != undefined && step.y != undefined){
      ellipse(step.x, step.y, 20, 20);
      fill(0);
      text(key, step.x-5, step.y+5);
    }
  });

  textSize(20)
  text(sequence, width/2-20, 20);
}